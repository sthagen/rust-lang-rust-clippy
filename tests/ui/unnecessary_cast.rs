//@run-rustfix
#![warn(clippy::unnecessary_cast)]
#![allow(
    unused,
    clippy::borrow_as_ptr,
    clippy::no_effect,
    clippy::nonstandard_macro_braces,
    clippy::unnecessary_operation
)]

type PtrConstU8 = *const u8;
type PtrMutU8 = *mut u8;

fn owo<T>(ptr: *const T) -> *const T {
    ptr as *const T
}

fn uwu<T, U>(ptr: *const T) -> *const U {
    ptr as *const U
}

#[rustfmt::skip]
fn main() {
    // Test cast_unnecessary
    1i32 as i32;
    1f32 as f32;
    false as bool;
    &1i32 as &i32;

    -1_i32 as i32;
    - 1_i32 as i32;
    -1f32 as f32;
    1_i32 as i32;
    1_f32 as f32;

    let _: *mut u8 = [1u8, 2].as_ptr() as *const u8 as *mut u8;

    [1u8, 2].as_ptr() as *const u8;
    [1u8, 2].as_ptr() as *mut u8;
    [1u8, 2].as_mut_ptr() as *mut u8;
    [1u8, 2].as_mut_ptr() as *const u8;
    [1u8, 2].as_ptr() as PtrConstU8;
    [1u8, 2].as_ptr() as PtrMutU8;
    [1u8, 2].as_mut_ptr() as PtrMutU8;
    [1u8, 2].as_mut_ptr() as PtrConstU8;
    let _: *const u8 = [1u8, 2].as_ptr() as _;
    let _: *mut u8 = [1u8, 2].as_mut_ptr() as _;
    let _: *const u8 = [1u8, 2].as_ptr() as *const _;
    let _: *mut u8 = [1u8, 2].as_mut_ptr() as *mut _;

    owo::<u32>([1u32].as_ptr()) as *const u32;
    uwu::<u32, u8>([1u32].as_ptr()) as *const u8;
    // this will not lint in the function body even though they have the same type, instead here
    uwu::<u32, u32>([1u32].as_ptr()) as *const u32;

    // macro version
    macro_rules! foo {
        ($a:ident, $b:ident) => {
            #[allow(unused)]
            pub fn $a() -> $b {
                1 as $b
            }
        };
    }
    foo!(a, i32);
    foo!(b, f32);
    foo!(c, f64);

    // do not lint cast to cfg-dependant type
    1 as std::os::raw::c_char;

    // do not lint cast to alias type
    1 as I32Alias;
    &1 as &I32Alias;

    let i8_ptr: *const i8 = &1;
    let u8_ptr: *const u8 = &1;

    // cfg dependant pointees
    i8_ptr as *const std::os::raw::c_char;
    u8_ptr as *const std::os::raw::c_char;

    // type aliased pointees
    i8_ptr as *const std::ffi::c_char;
    u8_ptr as *const std::ffi::c_char;

    // issue #9960
    macro_rules! bind_var {
        ($id:ident, $e:expr) => {{
            let $id = 0usize;
            let _ = $e != 0usize;
            let $id = 0isize;
            let _ = $e != 0usize;
        }}
    }
    bind_var!(x, (x as usize) + 1);
}

type I32Alias = i32;

mod fixable {
    #![allow(dead_code)]

    fn main() {
        // casting integer literal to float is unnecessary
        100 as f32;
        100 as f64;
        100_i32 as f64;
        let _ = -100 as f32;
        let _ = -100 as f64;
        let _ = -100_i32 as f64;
        100. as f32;
        100. as f64;
        // Should not trigger
        #[rustfmt::skip]
        let v = vec!(1);
        &v as &[i32];
        0x10 as f32;
        0o10 as f32;
        0b10 as f32;
        0x11 as f64;
        0o11 as f64;
        0b11 as f64;

        1 as u32;
        0x10 as i32;
        0b10 as usize;
        0o73 as u16;
        1_000_000_000 as u32;

        1.0 as f64;
        0.5 as f32;

        1.0 as u16;

        let _ = -1 as i32;
        let _ = -1.0 as f32;

        let _ = 1 as I32Alias;
        let _ = &1 as &I32Alias;

        let x = 1i32;
        let _ = &(x as i32);
    }

    type I32Alias = i32;

    fn issue_9380() {
        let _: i32 = -(1) as i32;
        let _: f32 = -(1) as f32;
        let _: i64 = -(1) as i64;
        let _: i64 = -(1.0) as i64;

        let _ = -(1 + 1) as i64;
    }

    fn issue_9563() {
        let _: f64 = (-8.0 as f64).exp();
        #[allow(clippy::precedence)]
        let _: f64 = -(8.0 as f64).exp(); // should suggest `-8.0_f64.exp()` here not to change code behavior
    }

    fn issue_9562_non_literal() {
        fn foo() -> f32 {
            0.
        }

        let _num = foo() as f32;
    }

    fn issue_9603() {
        let _: f32 = -0x400 as f32;
    }
}
