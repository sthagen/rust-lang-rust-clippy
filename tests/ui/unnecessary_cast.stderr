error: casting raw pointers to the same type and constness is unnecessary (`*const T` -> `*const T`)
  --> $DIR/unnecessary_cast.rs:15:5
   |
LL |     ptr as *const T
   |     ^^^^^^^^^^^^^^^ help: try: `ptr`
   |
   = note: `-D clippy::unnecessary-cast` implied by `-D warnings`

error: casting integer literal to `i32` is unnecessary
  --> $DIR/unnecessary_cast.rs:25:5
   |
LL |     1i32 as i32;
   |     ^^^^^^^^^^^ help: try: `1_i32`

error: casting float literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:26:5
   |
LL |     1f32 as f32;
   |     ^^^^^^^^^^^ help: try: `1_f32`

error: casting to the same type is unnecessary (`bool` -> `bool`)
  --> $DIR/unnecessary_cast.rs:27:5
   |
LL |     false as bool;
   |     ^^^^^^^^^^^^^ help: try: `false`

error: casting integer literal to `i32` is unnecessary
  --> $DIR/unnecessary_cast.rs:30:5
   |
LL |     -1_i32 as i32;
   |     ^^^^^^^^^^^^^ help: try: `-1_i32`

error: casting integer literal to `i32` is unnecessary
  --> $DIR/unnecessary_cast.rs:31:5
   |
LL |     - 1_i32 as i32;
   |     ^^^^^^^^^^^^^^ help: try: `- 1_i32`

error: casting float literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:32:5
   |
LL |     -1f32 as f32;
   |     ^^^^^^^^^^^^ help: try: `-1_f32`

error: casting integer literal to `i32` is unnecessary
  --> $DIR/unnecessary_cast.rs:33:5
   |
LL |     1_i32 as i32;
   |     ^^^^^^^^^^^^ help: try: `1_i32`

error: casting float literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:34:5
   |
LL |     1_f32 as f32;
   |     ^^^^^^^^^^^^ help: try: `1_f32`

error: casting raw pointers to the same type and constness is unnecessary (`*const u8` -> `*const u8`)
  --> $DIR/unnecessary_cast.rs:36:22
   |
LL |     let _: *mut u8 = [1u8, 2].as_ptr() as *const u8 as *mut u8;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `[1u8, 2].as_ptr()`

error: casting raw pointers to the same type and constness is unnecessary (`*const u8` -> `*const u8`)
  --> $DIR/unnecessary_cast.rs:38:5
   |
LL |     [1u8, 2].as_ptr() as *const u8;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `[1u8, 2].as_ptr()`

error: casting raw pointers to the same type and constness is unnecessary (`*mut u8` -> `*mut u8`)
  --> $DIR/unnecessary_cast.rs:40:5
   |
LL |     [1u8, 2].as_mut_ptr() as *mut u8;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `[1u8, 2].as_mut_ptr()`

error: casting raw pointers to the same type and constness is unnecessary (`*const u32` -> `*const u32`)
  --> $DIR/unnecessary_cast.rs:49:5
   |
LL |     owo::<u32>([1u32].as_ptr()) as *const u32;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `owo::<u32>([1u32].as_ptr())`

error: casting raw pointers to the same type and constness is unnecessary (`*const u8` -> `*const u8`)
  --> $DIR/unnecessary_cast.rs:50:5
   |
LL |     uwu::<u32, u8>([1u32].as_ptr()) as *const u8;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `uwu::<u32, u8>([1u32].as_ptr())`

error: casting raw pointers to the same type and constness is unnecessary (`*const u32` -> `*const u32`)
  --> $DIR/unnecessary_cast.rs:52:5
   |
LL |     uwu::<u32, u32>([1u32].as_ptr()) as *const u32;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `uwu::<u32, u32>([1u32].as_ptr())`

error: casting integer literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:93:9
   |
LL |         100 as f32;
   |         ^^^^^^^^^^ help: try: `100_f32`

error: casting integer literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:94:9
   |
LL |         100 as f64;
   |         ^^^^^^^^^^ help: try: `100_f64`

error: casting integer literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:95:9
   |
LL |         100_i32 as f64;
   |         ^^^^^^^^^^^^^^ help: try: `100_f64`

error: casting integer literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:96:17
   |
LL |         let _ = -100 as f32;
   |                 ^^^^^^^^^^^ help: try: `-100_f32`

error: casting integer literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:97:17
   |
LL |         let _ = -100 as f64;
   |                 ^^^^^^^^^^^ help: try: `-100_f64`

error: casting integer literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:98:17
   |
LL |         let _ = -100_i32 as f64;
   |                 ^^^^^^^^^^^^^^^ help: try: `-100_f64`

error: casting float literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:99:9
   |
LL |         100. as f32;
   |         ^^^^^^^^^^^ help: try: `100_f32`

error: casting float literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:100:9
   |
LL |         100. as f64;
   |         ^^^^^^^^^^^ help: try: `100_f64`

error: casting integer literal to `u32` is unnecessary
  --> $DIR/unnecessary_cast.rs:112:9
   |
LL |         1 as u32;
   |         ^^^^^^^^ help: try: `1_u32`

error: casting integer literal to `i32` is unnecessary
  --> $DIR/unnecessary_cast.rs:113:9
   |
LL |         0x10 as i32;
   |         ^^^^^^^^^^^ help: try: `0x10_i32`

error: casting integer literal to `usize` is unnecessary
  --> $DIR/unnecessary_cast.rs:114:9
   |
LL |         0b10 as usize;
   |         ^^^^^^^^^^^^^ help: try: `0b10_usize`

error: casting integer literal to `u16` is unnecessary
  --> $DIR/unnecessary_cast.rs:115:9
   |
LL |         0o73 as u16;
   |         ^^^^^^^^^^^ help: try: `0o73_u16`

error: casting integer literal to `u32` is unnecessary
  --> $DIR/unnecessary_cast.rs:116:9
   |
LL |         1_000_000_000 as u32;
   |         ^^^^^^^^^^^^^^^^^^^^ help: try: `1_000_000_000_u32`

error: casting float literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:118:9
   |
LL |         1.0 as f64;
   |         ^^^^^^^^^^ help: try: `1.0_f64`

error: casting float literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:119:9
   |
LL |         0.5 as f32;
   |         ^^^^^^^^^^ help: try: `0.5_f32`

error: casting integer literal to `i32` is unnecessary
  --> $DIR/unnecessary_cast.rs:123:17
   |
LL |         let _ = -1 as i32;
   |                 ^^^^^^^^^ help: try: `-1_i32`

error: casting float literal to `f32` is unnecessary
  --> $DIR/unnecessary_cast.rs:124:17
   |
LL |         let _ = -1.0 as f32;
   |                 ^^^^^^^^^^^ help: try: `-1.0_f32`

error: casting to the same type is unnecessary (`i32` -> `i32`)
  --> $DIR/unnecessary_cast.rs:130:18
   |
LL |         let _ = &(x as i32);
   |                  ^^^^^^^^^^ help: try: `{ x }`

error: casting integer literal to `i32` is unnecessary
  --> $DIR/unnecessary_cast.rs:136:22
   |
LL |         let _: i32 = -(1) as i32;
   |                      ^^^^^^^^^^^ help: try: `-1_i32`

error: casting integer literal to `i64` is unnecessary
  --> $DIR/unnecessary_cast.rs:138:22
   |
LL |         let _: i64 = -(1) as i64;
   |                      ^^^^^^^^^^^ help: try: `-1_i64`

error: casting float literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:145:22
   |
LL |         let _: f64 = (-8.0 as f64).exp();
   |                      ^^^^^^^^^^^^^ help: try: `(-8.0_f64)`

error: casting float literal to `f64` is unnecessary
  --> $DIR/unnecessary_cast.rs:147:23
   |
LL |         let _: f64 = -(8.0 as f64).exp(); // should suggest `-8.0_f64.exp()` here not to change code behavior
   |                       ^^^^^^^^^^^^ help: try: `8.0_f64`

error: casting to the same type is unnecessary (`f32` -> `f32`)
  --> $DIR/unnecessary_cast.rs:155:20
   |
LL |         let _num = foo() as f32;
   |                    ^^^^^^^^^^^^ help: try: `foo()`

error: aborting due to 38 previous errors

